program-cache/lib/src/program_cache.cpp

```diff
@@ -0,0 +1,321 @@
+                   L"LOCALAPPDATA"))
+    {
+        throw cache_access_error("Could not get default cache root directory");
+    }
+    else
+    {
+        appdata_local.resize(buffer_length);
+        return std::filesystem::path(appdata_local) / "Khronos" / "OpenCL"
+            / "cache";
+    }
+}
+
+#elif defined(__APPLE__)
+
+std::filesystem::path get_default_cache_root()
+{
+    if (char* home_path = std::getenv("HOME"); home_path == nullptr)
+    {
+        throw cache_access_error("Could not get default cache root directory");
+    }
+    else
+    {
+        return std::filesystem::path(home_path) / "Library" / "Caches"
+            / "Khronos" / "OpenCL";
+    }
+}
+
+#else
+
+std::filesystem::path get_default_cache_root()
+{
+    const auto cache_home = []() -> std::filesystem::path {
+        if (char* cache_home = std::getenv("XDG_CACHE_HOME");
+            cache_home == nullptr)
+        {
+            if (char* home_path = std::getenv("HOME"); home_path == nullptr)
+            {
+                throw cache_access_error(
+                    "Could not get default cache root directory");
+            }
+            else
+            {
+                return std::filesystem::path(home_path) / ".cache";
+            }
+        }
+        else
+        {
+            return { cache_home };
+        }
+    }();
+    return cache_home / "Khronos" / "OpenCL" / "cache";
+}
+
+#endif
+
+std::string get_device_identifier(const cl::Device& device)
+{
+    const auto device_name = device.getInfo<CL_DEVICE_NAME>();
+    const auto platform = cl::Platform(device.getInfo<CL_DEVICE_PLATFORM>());
+    const auto platform_version = platform.getInfo<CL_PLATFORM_VERSION>();
+    return platform_version + "/" + device_name;
+}
+
+std::vector<unsigned char> build_program_to_binary(const cl::Context& context,
+                                                   const cl::Device& device,
+                                                   std::string_view source,
+                                                   std::string_view options)
+{
+    cl::Program program(context, std::string(source));
+    CHECK_CL_BUILD_ERROR(program.build(device, options.data()));
+    return program.getInfo<CL_PROGRAM_BINARIES>().front();
+}
+
+std::vector<unsigned char> build_program_to_binary(const cl::Context& context,
+                                                   const cl::Device& device,
+                                                   const std::vector<char>& il,
+                                                   std::string_view options)
+{
+    cl::Program program(context, il);
+    CHECK_CL_BUILD_ERROR(program.build(device, options.data()));
+    return program.getInfo<CL_PROGRAM_BINARIES>().front();
+}
+
+void save_binary_to_cache(const std::filesystem::path& cache_path,
+                          const std::vector<unsigned char>& binary_data)
+{
+    std::filesystem::create_directory(cache_path.parent_path());
```

Comment:
Perhaps
```suggestion
    std::filesystem::create_directories(cache_path.parent_path());
```
?

As I saw, the binaries are stored in `cache_root/2FirstKeyHashChars/RestOfKeyHash_DeviceHash`, so the intermediate directory between the `cache_root` and the `RestOfKeyHash_DeviceHash` directory does not need to exist the first time the files are cached?

However, the tests executed in the GitHub actions pass, so am I missing something?

Reply:
Cache root is created in the constructor with `create_directories`. And `2FirstKeyHashChars` is created by this line (`cache_path.parent_path()` stands for `cache_root/2FirstKeyHashChars`). So I think this is correct as is.